### PR TITLE
Accessibility behavior tests and documentation generation

### PR DESCRIPTION
**Pull request by kolaps33**
#### Pull request checklist

- [ ] Include a change request file using `$ yarn change`

#### Open questions/Future work
- invoke click and keyboard event with testing libary - https://github.com/testing-library/user-event > can be implement in future after agreement inside the team 
- where should be "[behaviorName]Definition" files stored? so far the definition file is in the same folder where is "behavior" file 
- currently generating documentation from "behavior" file and from "definition" exists together. 
If documentation is not found in behavior file, then is generated from "definition" file. In the future generating doc from behavior file need to be removed. 

#### Description of changes
This PR is coming out from the @mnajdova PR. 
https://github.com/microsoft/fluentui/pull/14797

Purpose of the PR is to test "behavior" files and generate documentation from each behavior file based on the test. 

In this PR following behaviors are tested:
- buttonBehavior
- buttonGroupBehavior
- toggleButtonBehavior

This behaviors are now cover by test which stpes are in the "[behaviorName]Definition" file. 

From "[behaviorName]Definition" file is as well documentation for each behavior file generated. 

It doesn't change the documentation, only content under "description" paragraph is coming from "[behaviorName]Definition" file (pl. see screenshot below):
image

#### Focus areas to test

(optional)

### CODE REVIEW

File: packages/fluentui/react-northstar/src/components/Button/Button.tsx [hunk 1]
```diff
@@ -232,7 +232,6 @@ export const Button = compose<'button', ButtonProps, ButtonStylesProps, {}, {}>(
         {...rtlTextContainer.getAttributes({ forElements: [children] })}
         {...getA11yProps('root', {
           onClick: handleClick,
-          disabled,
```
Comment (kolaps33) [inline packages/fluentui/react-northstar/src/components/Button/Button.tsx:235]: Removing this line because it break the condition

` disabled: props.disabled && !props.loading ? (props.as === 'button' ? true : undefined) : undefined,`

which is in ButtonBehavior.ts file.

File: packages/react-button/src/components/ToggleButton/useChecked.test.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { useChecked } from './useChecked';
+import { checkedBehaviorDefinition } from './checkedBehaviorDefinition';
+import { HookTestFacade, validateBehavior } from '@fluentui/a11y-testing';
+
+describe('useChecked', () => {
```
Comment (mnajdova) [inline packages/react-button/src/components/ToggleButton/useChecked.test.ts:5]: Let's remove this file, it's not relevant anymore

File: packages/react-button/src/components/ToggleButton/checkedBehaviorDefinition.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { BehaviorRule, Rule } from '@fluentui/a11y-testing';
```
Comment (mnajdova) [inline packages/react-button/src/components/ToggleButton/checkedBehaviorDefinition.ts:1]: Let's remove this file, it's not relevant anymore

File: packages/react-button/src/components/ToggleButton/ToggleButton.test.tsx [hunk 4]
```diff
@@ -1,6 +1,8 @@
 import * as path from 'path';
 import { isConformant } from '@fluentui/react-conformance';
 import { ToggleButton } from './ToggleButton';
+import { checkedBehaviorDefinition } from './checkedBehaviorDefinition';
```
Comment (mnajdova) [inline packages/react-button/src/components/ToggleButton/ToggleButton.test.tsx:4]: Let's revert the changes here too, not relevant anymore

File: packages/react-button/package.json [hunk 5]
```diff
@@ -50,6 +50,7 @@
     "react-test-renderer": "^16.3.0"
   },
   "dependencies": {
+    "@fluentui/a11y-testing": "^0.1.0",
```
Comment (mnajdova) [inline packages/react-button/package.json:53]:
```suggestion
```

File: packages/react-avatar/.eslint.json [hunk 6]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (mnajdova) [inline packages/react-avatar/.eslint.json:1]: Check if this is still required, I was getting some error so I needed to do this change

File: packages/charting/src/components/VerticalStackedBarChart/examples/VerticalStackedBarChart.Styled.Example.tsx [hunk 7]
```diff
@@ -76,15 +76,15 @@ export class VerticalStackedBarChartStyledExample extends React.Component<{}, {}
             allowFocusOnLegends: true,
             styles: {
               rect: {
-                borderRadius: "3px"
-              }
-            }
+                borderRadius: '3px',
```
Comment (mnajdova) [inline packages/charting/src/components/VerticalStackedBarChart/examples/VerticalStackedBarChart.Styled.Example.tsx:79]: Let's revert these changes, they are not relevant anyhow to this PR.

File: packages/charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx [hunk 8]
```diff
@@ -148,13 +148,11 @@ export class VerticalStackedBarChartBase extends React.Component<
...
           {...this.props.calloutProps}
         >
-          {
-            this.props.onRenderCalloutPerStack
-              ? this.props.onRenderCalloutPerStack(this.state.stackCalloutProps)
-              : this.props.onRenderCalloutPerDataPoint
-                ? this.props.onRenderCalloutPerDataPoint(this.state.dataPointCalloutProps, this._renderCallout)
-                : this._renderCallout(this.state.dataPointCalloutProps)
-          }
+          {this.props.onRenderCalloutPerStack
```
Comment (mnajdova) [inline packages/charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:151]: Let's revert these changes too

File: packages/a11y-testing/src/validators/validate.ts [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+    if (slot.checkSpaceKeyPressed) {
+      testFacade.pressSpaceKey(slot.name);
+    }
+
+    if (slot.checkEnterKeyPressed) {
+      testFacade.pressEnterKey(slot.name);
+    }
+
+    if (slot.wasOnclickExecuted) {
```
Comment (mnajdova) [inline packages/a11y-testing/src/validators/validate.ts:23]: For consistensy, can we rename this to checkOnClickWasExecuted, or something similar

File: packages/a11y-testing/src/facades/withProvider.tsx [hunk 10]
```diff
@@ -0,0 +1,43 @@
+import { Telemetry, Unstable_FluentContextProvider, ProviderContextPrepared } from '@fluentui/react-bindings';
+import { Renderer, noopRenderer } from '@fluentui/react-northstar-styles-renderer';
+import { emptyTheme, ThemePrepared } from '@fluentui/styles';
+import { mount, MountRendererProps, ComponentType } from 'enzyme';
+import * as React from 'react';
+
+export const EmptyThemeProvider: React.FunctionComponent<{
```
Comment (mnajdova) [inline packages/a11y-testing/src/facades/withProvider.tsx:7]: @kolaps33 can we try to test if everything will work if we just don't wrap the component with theme provider. This was initially implemented like this, because the components will throw an error, but I am not sure what is the status at this moment. Would be great if we can remove the styles and other dependencies from this package.

File: packages/a11y-testing/src/facades/ComponentTestFacade.tsx [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    // reset body
+    document.body.innerHTML = '';
+    const container = document.createElement('div');
+    document.body.appendChild(container);
+
+    ReactDOM.render(<Component {...props} />, container);
+    this.actual = container.lastChild;
+
+    // we need to render it in this way because using simulate function to fire mouse/keyboard event
+    this.renderedComponent = mountWithProviderAndGetComponent(Component, <Component {...props} />);
```
Comment (mnajdova) [inline packages/a11y-testing/src/facades/ComponentTestFacade.tsx:25]: I was refering to trying to remove this `mountWithProviderAndGetComponent` and just render the `Component` itself.

File: packages/a11y-testing/package.json [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    "test": "just-scripts test",
+    "update-api": "just-scripts update-api",
+    "update-snapshots": "just-scripts jest -u"
+  },
+  "devDependencies": {
+    "@fluentui/common-styles": "^0.2.2",
+    "@fluentui/eslint-plugin": "^0.53.3",
+    "@fluentui/react-conformance": "^0.1.0",
+    "@fluentui/storybook": "^0.2.6",
+    "@testing-library/react": "^10.4.9",
```
Comment (mnajdova) [inline packages/a11y-testing/package.json:34]: I don't think we are using this anymore, as far as I remember it was only used in the `HooksTestFacade`, but we decided to go only with the `ComponentsTestFacade`

**Review by mnajdova — COMMENTED** [review]
Let's add comments on the methods, to document what they are used for. It would help people get up to speed faster

File: packages/a11y-testing/etc/a11y-testing.api.md [hunk 13]
```diff
@@ -0,0 +1,146 @@
+## API Report File for "@fluentui/a11y-testing"
```
Comment (kolaps33) [inline packages/a11y-testing/etc/a11y-testing.api.md:1]: This file was generating by running command in the package folder:
yarn update-api

and then manually copied from temp folder to this etc folder

File: packages/a11y-testing/package.json (hunk 12 above)
Comment (kolaps33) [reply packages/a11y-testing/package.json]: removed

File: packages/charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx (hunk 8 above)
Comment (kolaps33) [reply packages/charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx:151]: somehow commit always change formatting this file... I don't know how to fix it.

File: packages/charting/src/components/VerticalStackedBarChart/examples/VerticalStackedBarChart.Styled.Example.tsx (hunk 7 above)
Comment (kolaps33) [reply packages/charting/src/components/VerticalStackedBarChart/examples/VerticalStackedBarChart.Styled.Example.tsx:79]: somehow commit always change formatting this file... I don't know how to fix it.

File: packages/a11y-testing/etc/a11y-testing.api.md (hunk 13 above)
Comment (ecraig12345) [reply packages/a11y-testing/etc/a11y-testing.api.md:1]: `build` will give you the message about manually copying, but `update-api` should copy automatically. Also if this isn't intended to be a published package we can just turn off API Extractor.

File: packages/a11y-testing/package.json [hunk 14]
```diff
@@ -0,0 +1,61 @@
+{
+  "name": "@fluentui/a11y-testing",
+  "version": "0.1.0",
```
Comment (ecraig12345) [inline packages/a11y-testing/package.json:3]: Is this intended to be published? If not you can add `"private": true`

File: packages/a11y-testing/package.json [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+  "description": "Fluent UI tests for a11y patterns.",
+  "main": "lib-commonjs/index.js",
```
Comment (ecraig12345) [inline packages/a11y-testing/package.json:5]: From what I can tell, this package should have used the `node` template not the `react` template since it's intended to be used for testing only, not run in the browser (though I agree the template name is a little misleading). So now there's a bunch of configuration that's not quite right. 

Is it okay if I push an update to your PR fixing the config issues? (would be easier than trying to explain them all)

File: packages/react-button/src/components/ToggleButton/useChecked.test.ts (hunk 2 above)
Comment (kolaps33) [reply packages/react-button/src/components/ToggleButton/useChecked.test.ts]: removed :)

File: packages/react-button/src/components/ToggleButton/checkedBehaviorDefinition.ts (hunk 3 above)
Comment (kolaps33) [reply packages/react-button/src/components/ToggleButton/checkedBehaviorDefinition.ts]: removed :)

File: packages/react-button/src/components/ToggleButton/ToggleButton.test.tsx (hunk 4 above)
Comment (kolaps33) [reply packages/react-button/src/components/ToggleButton/ToggleButton.test.tsx]: changes reverted

File: packages/react-button/package.json (hunk 5 above)
Comment (kolaps33) [reply packages/react-button/package.json]: removed :)

File: packages/react-avatar/.eslint.json (hunk 6 above)
Comment (kolaps33) [reply packages/react-avatar/.eslint.json]: removed whole file

File: packages/a11y-testing/src/validators/validate.ts (hunk 9 above)
Comment (kolaps33) [reply packages/a11y-testing/src/validators/validate.ts]: renamed :)

File: packages/a11y-testing/src/facades/withProvider.tsx (hunk 10 above)
Comment (kolaps33) [reply packages/a11y-testing/src/facades/withProvider.tsx]: removed whole file and replace it with one line in `ComponentTestFacade.tsx`

File: packages/a11y-testing/src/facades/ComponentTestFacade.tsx (hunk 11 above)
Comment (kolaps33) [reply packages/a11y-testing/src/facades/ComponentTestFacade.tsx]: yes thanks, replaced with line:
` this.renderedComponent = mount(<Component {...props} />).find(Component);`

File: packages/fluentui/react-northstar/test/specs/components/Button/ButtonBehavior-test.tsx [hunk 16]
```diff
@@ -0,0 +1,15 @@
+// import * as path from 'path';
```
[the remark this answers is not in the document]
Comment (kolaps33) [reply packages/fluentui/react-northstar/test/specs/components/Button/ButtonBehavior-test.tsx:1]: agree, I am wondering what is opinion of @jurokapsiar

File: packages/a11y-testing/package.json (hunk 15 above)
Comment (kolaps33) [reply packages/a11y-testing/package.json:5]: @ecraig12345 
please feel free to push update to this PR :) 
I appreciate you help. :) as I have no experience with creating the package. 
Even if you have any documentation where I could read more about creating package I would be happy to read it.

File: packages/a11y-testing/package.json (hunk 14 above)
Comment (kolaps33) [reply packages/a11y-testing/package.json:3]: I am not sure about long term plan. @jurokapsiar please could you clarify? :)

Comment (jurokapsiar) [reply packages/a11y-testing/package.json:3]: not now, just internal. maybe in the future

File: packages/a11y-testing/package.json (hunk 15 above)
Comment (ecraig12345) [reply packages/a11y-testing/package.json:5]: It's not documented 😬 but that would be a good idea since we're adding more packages now! I'm also planning to update the create-package script with more options (and clearer explanations) so it's easy for people to get the right setup in more cases.

File: scripts/gulp/plugins/gulp-component-menu-behaviors.ts [hunk 17]
```diff
@@ -20,6 +20,11 @@ type BehaviorMenuItem = {
   };
 };
 
+async function importFile(pathDir, name) {
+  const path = `${pathDir}\\${name}.ts`;
```
Comment (ecraig12345) [inline scripts/gulp/plugins/gulp-component-menu-behaviors.ts:24]: You should probably use `path.join` here since backslashes are Windows-specific

**Comment by ecraig12345** [comment]
I just pushed an update which does the following:
- Set package to `private: true` and remove change file (we can always publish it later if needed)
- Change eslint config to extend the one for node (build/test utility) packages
- Removed storybook since it's only used for component packages
- Removed API Extractor config (locking down API isn't important for an internal package)
- Removed test configs (and most test-related dependencies): although this package has test-related utilities, it doesn't appear to have its own tests. If that changes, the test-related config can be added back.
- Change just.config.ts to use node library-style build (so it will only output commonjs modules to `lib`, rather than the component package build where ES modules go in `lib` and commonjs modules go in `lib-commonjs`) and updated entry points in package.json
- Remove unused scripts and dependencies
- Update tsconfig to use options for a node library: mainly removing unneeded types and using `lib: [es2017, dom]` instead of `es5` (for browser packages we restrict to ES5 types for IE 11 compatibility, but node libraries can use newer types)

File: packages/charting/src/components/VerticalStackedBarChart/examples/VerticalStackedBarChart.Styled.Example.tsx (hunk 7 above)
Comment (kolaps33) [reply packages/charting/src/components/VerticalStackedBarChart/examples/VerticalStackedBarChart.Styled.Example.tsx]: with merge latest master it get fixed

File: packages/charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx (hunk 8 above)
Comment (kolaps33) [reply packages/charting/src/components/VerticalStackedBarChart/VerticalStackedBarChart.base.tsx]: with merge latest master it get fixed

File: packages/fluentui/react-northstar/test/specs/components/Button/ButtonBehavior-test.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (kolaps33) [reply packages/fluentui/react-northstar/test/specs/components/Button/ButtonBehavior-test.tsx]: change done, tests were moved into the common component tests to be together in one file

File: scripts/gulp/plugins/gulp-component-menu-behaviors.ts (hunk 17 above)
Comment (kolaps33) [reply scripts/gulp/plugins/gulp-component-menu-behaviors.ts]: thanks, implemented proposal change

File: packages/a11y-testing/package.json [hunk 18]
```diff
@@ -18,6 +18,7 @@
     "lint": "just-scripts lint"
   },
   "dependencies": {
+    "@fluentui/a11y-rules": "0.1.0",
```
Comment (layershifter) [inline packages/a11y-testing/package.json:21]:
```suggestion
    "@fluentui/a11y-rules": "^0.1.0",
```

File: packages/fluentui/accessibility/package.json [hunk 19]
```diff
@@ -10,6 +10,7 @@
     "lodash": "^4.17.15"
   },
   "devDependencies": {
+    "@fluentui/a11y-rules": "0.1.0",
     "@fluentui/a11y-testing": "0.1.0",
```
Comment (layershifter) [inline packages/fluentui/accessibility/package.json]:
```suggestion
    "@fluentui/a11y-rules": "^0.1.0",
    "@fluentui/a11y-testing": "^0.1.0",
```

File: packages/fluentui/accessibility-definitions/package.json [hunk 20]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "@fluentui/accessibility-definitions",
+  "description": "A set of acc.... ",
```
Comment (ecraig12345) [inline packages/fluentui/accessibility-definitions/package.json:3]: This looks unfinished? Also I'm curious why another package ended up being needed.

Comment (kolaps33) [reply packages/fluentui/accessibility-definitions/package.json]: @ecraig12345  thank you for cath. Updated. 

The new package I am adding because of dependency which "definitions" files have inside. 
The "definitions" files depends on "a11y-rules" package. 

"Definitions" files were before together with "behaviors" files in "accessibility" package:
office-ui-fabric-react\packages\fluentui\accessibility\src\behaviors

"Accessibility" package is public.
But we don't want to make "a11y-rules" package public. (as dependency of "definittions" files) 

So this was the reason extract "definitions" files out of the "accessibility" package.

File: packages/fluentui/accessibility-definitions/src/definitions/Button/buttonBehaviorDefinition.ts [hunk 21]
```diff
@@ -0,0 +1,48 @@
+import { Rule, BehaviorRule } from '@fluentui/a11y-rules';
```
Comment (layershifter) [inline packages/fluentui/accessibility-definitions/src/definitions/Button/buttonBehaviorDefinition.ts:1]: I suggest to eliminate there a level of nesting: `fluentui/accessibility-definitions/src/definitions/Button/buttonBehaviorDefinition.ts` => `fluentui/accessibility-definitions/src/Button/buttonBehaviorDefinition.ts`

Comment (layershifter) [reply packages/fluentui/accessibility-definitions/src/definitions/Button/buttonBehaviorDefinition.ts]: I did this change

File: scripts/gulp/plugins/gulp-component-menu-behaviors.ts [hunk 22]
```diff
@@ -59,12 +74,32 @@ export default () => {
...
+            return definition.stringify();
+          });
+          variation.description = descriptionFromDefinition.join('\r\n');
+          result.push({
+            displayName: behaviorName,
+            type: componentType,
+            variations: variation,
+          });
+          cb();
+        });
```
Comment (layershifter) [inline scripts/gulp/plugins/gulp-component-menu-behaviors.ts:87]: I simplified this part, please check that it still works

**Review by layershifter — APPROVED** [review]

File: scripts/gulp/plugins/gulp-component-menu-behaviors.ts (hunk 22 above)
Comment (kolaps33) [reply scripts/gulp/plugins/gulp-component-menu-behaviors.ts:71]: thank you very much, verified, tested, works as expected

File: packages/fluentui/accessibility-definitions/package.json (hunk 20 above)
Comment (ecraig12345) [reply packages/fluentui/accessibility-definitions/package.json]: I'm still a little confused. If the definitions are only needed as a **devDependency** of `accessibility`, they don't need to be in a published package. But if they're going to be used (directly or indirectly) by consumers of `accessibility`, the packages need to be public. Also, you can't have a private package as (non-dev) dependency of a published package, even indirectly.

Comment (kolaps33) [reply packages/fluentui/accessibility-definitions/package.json]: I think "definitions" we don't want to be consume/used with accessibility package. 

Because this definitions are used only for testing of "Behaviors" files and user of "accessibility" can use only "behaviors" files. 

`Also, you can't have a private package as (non-dev) dependency of a published package, even indirectly.`
if this issue, should I revert changes and put "definitions" package back to the "rules" package and make "rules" package public?
and if it will be public do I need create separate PR for it? :)

Comment (miroslavstastny) [reply packages/fluentui/accessibility-definitions/package.json]: We discussed this offline, the distinction among `a11y-testing`, `a11y-rules` and `accessibility-definitions` is hard to explain which probably means it does not have much sense. Based on that we agreed to have the two following packages:
- `@fluentui/accessibility` - PUBLIC package, contains behaviors but not their description.
- `@fluentui/a11y-testing` - private package, contains behavior descriptions, testing rules and testing framework for it.

**Comment by ecraig12345** [comment]
I think the build failure is because you need to add an alias in a tsconfig somewhere for `@fluentui/a11y-testing` since it's not within `packages/fluentui`. Try adding this under `compilerOptions.paths` in `packages/fluentui/react-northstar/tsconfig.json`: `"@fluentui/a11y-testing": ["packages/a11y-testing/src"]` (and if that doesn't work I'd recommend asking Shift or Miro if they can help figure out where to add it)

**Review by ecraig12345 — APPROVED** [review]

**Review by jurokapsiar — APPROVED** [review]
